File: examples/05-durable-workflow/handlers.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Workflow step handlers for the durable-workflow example.

The worker process imports this module on startup. The decorations
populate ``client._workflow_runtime`` so the worker's dispatch table
matches the manifest the start script creates.

Workflow:  ``research-pipeline``
Steps:     ``search`` → ``fetch`` → ``extract`` → ``synth``

Each step writes its outputs to the workspace KV / files so the next
step can read them. Snapshots are taken at every boundary so a
crash + worker restart can resume from the last known good state
without redoing work.
"""

from __future__ import annotations

import time

from plinth import Plinth

from shared import (
    make_client_kwargs,
    mock_extract,
    mock_fetch,
    mock_search,
    mock_synthesise,
    slugify,
)

# ---------------------------------------------------------------------------
# Client + handler registrations
# ---------------------------------------------------------------------------
#
# Both this handlers module AND ``start_workflow.py`` construct a
# ``Plinth`` client; that's fine — they're separate processes. The worker
# loader (``__main__.py`` in plinth-workflow-worker) re-uses *this*
# module's client instance (and therefore its runtime registry) by
# scanning the imported module for ``Plinth`` attributes.

client = Plinth(**make_client_kwargs())


@client.workflow_handler("research-pipeline", step="search")
def search_step(ctx) -> dict:
    """Find ``k`` candidate sources for the input topic."""
    topic = ctx.step.input["topic"]
    sources = mock_search(topic, k=ctx.step.input.get("k", 5))
    ctx.workspace.kv.set("topic", topic)
    ctx.workspace.kv.set("sources/index", [s["url"] for s in sources])
    for s in sources:
        ctx.workspace.kv.set(f"sources/meta/{s['url']}", s)
    # Snapshot: worker will reference this snapshot in the step row so
    # a downstream resumer knows where to restore from.
    snap = ctx.workspace.snapshot(
        f"after-search-{int(time.time())}",
        message=f"search complete for {topic!r}",
    )
    return {"sources_count": len(sources), "snapshot_id": snap.id}


@client.workflow_handler("research-pipeline", step="fetch")
def fetch_step(ctx) -> dict:
    """Fetch each source's content and write to versioned files."""
    sources_idx = ctx.workspace.kv.get("sources/index")
    fetched = 0
    for url in sources_idx:
        content = mock_fetch(url)
        ctx.workspace.files.write(f"sources/{slugify(url)}.txt", content)
        fetched += 1
    snap = ctx.workspace.snapshot(
        f"after-fetch-{int(time.time())}",
        message=f"fetched {fetched} sources",
    )
    return {"fetched_count": fetched, "snapshot_id": snap.id}


@client.workflow_handler("research-pipeline", step="extract")
def extract_step(ctx) -> dict:
    """Extract facts from each fetched source."""
    topic = ctx.workspace.kv.get("topic")
    sources_idx = ctx.workspace.kv.get("sources/index")
    extracted = 0
    for url in sources_idx:
        path = f"sources/{slugify(url)}.txt"
        try:
            content = ctx.workspace.files.read(path, as_text=True)
        except Exception:  # noqa: BLE001 — file race; skip
            continue
        facts = mock_extract(content, topic=topic)
        ctx.workspace.kv.set(f"facts/{url}", facts)
        extracted += 1
    snap = ctx.workspace.snapshot(
        f"after-extract-{int(time.time())}",
        message=f"extracted from {extracted} sources",
    )
    return {"extracted_count": extracted, "snapshot_id": snap.id}


@client.workflow_handler("research-pipeline", step="synth")
def synth_step(ctx) -> dict:
    """Synthesise a final report from per-source facts."""
    topic = ctx.workspace.kv.get("topic")
    sources_idx = ctx.workspace.kv.get("sources/index")
    facts_by_url: dict[str, list[str]] = {}
    for url in sources_idx:
        try:
            facts_by_url[url] = ctx.workspace.kv.get(f"facts/{url}")
        except Exception:  # noqa: BLE001
            continue
    report = mock_synthesise(topic, facts_by_url)
    ctx.workspace.files.write("report.md", report)
    snap = ctx.workspace.snapshot(
        f"after-synth-{int(time.time())}",
        message="report written",
    )
    return {"report_chars": len(report), "snapshot_id": snap.id}


__all__ = ["client"]

File: examples/05-durable-workflow/handlers.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 The Plinth Authors
 *
 * Workflow step handlers — TypeScript counterpart of `handlers.py`.
 *
 * The TS worker (`@plinth/workflow-worker`) imports this module and
 * calls `register(runtime, client)` to populate the dispatch table. The
 * registered keys must match the manifest in `start-workflow.ts`:
 *
 *   workflow:  research-pipeline
 *   steps:     search → fetch → extract → synth
 *
 * Each step writes its outputs to the workspace KV / files so the next
 * step can read them; snapshots are taken at every boundary so a crash
 * + worker restart resumes from the last known good state.
 *
 * `extract` calls the LLM facade via `ctx.client.llm.complete(...)`. By
 * default the example wires `MockProvider` (offline, deterministic) so
 * the workflow runs without an API key — flip to Anthropic or OpenAI
 * in `start-workflow.ts` once you've exported the relevant env var.
 */

import { MockProvider, type Plinth, type JsonValue } from "@plinth/sdk";
import type { WorkflowRuntime } from "@plinth/workflow-worker";

import {
  mockFetch,
  mockSearch,
  mockSynthesise,
  slugify,
} from "./shared.js";

/**
 * Wire an LLM provider before the first handler runs.
 *
 *   * `ANTHROPIC_API_KEY` → AnthropicProvider (real Claude calls).
 *   * `OPENAI_API_KEY`   → OpenAIProvider (real GPT calls).
 *   * otherwise           → MockProvider (deterministic, offline).
 *
 * The auto-detection inside `LLMClient` would also resolve env keys at
 * the first `complete()` call, but configuring eagerly here makes the
 * worker boot fail loud — and gives us a clean place to install the
 * offline mock when no key is set.
 */
async function configureLLM(client: Plinth): Promise<void> {
  if (process.env.ANTHROPIC_API_KEY) {
    await client.llm.useProvider("anthropic", {
      apiKey: process.env.ANTHROPIC_API_KEY,
    });
    return;
  }
  if (process.env.OPENAI_API_KEY) {
    await client.llm.useProvider("openai", {
      apiKey: process.env.OPENAI_API_KEY,
    });
    return;
  }
  // Offline default — same shape and timing as a real provider, no
  // network calls. Useful for `make demo-durable-ts` runs without an
  // API key in the environment.
  client.llm.useCustomProvider(
    new MockProvider({
      responses: [
        "- Mock fact one\n- Mock fact two\n- Mock fact three",
      ],
      defaultModel: "claude-sonnet-4-5",
    }),
  );
}

interface SearchInput {
  topic: string;
  k?: number;
}

interface MockSource {
  url: string;
  title: string;
  snippet: string;
}

export async function register(runtime: WorkflowRuntime, client: Plinth): Promise<void> {
  await configureLLM(client);

  runtime.register("research-pipeline", "search", async (ctx) => {
    const input = ctx.step.input as unknown as SearchInput;
    const sources = mockSearch(input.topic, input.k ?? 5);
    await ctx.workspace.kv.set("topic", input.topic);
    await ctx.workspace.kv.set(
      "sources/index",
      sources.map((s: MockSource) => s.url),
    );
    for (const s of sources) {
      await ctx.workspace.kv.set(`sources/meta/${s.url}`, s as unknown as JsonValue);
    }
    const snap = await ctx.workspace.snapshot(
      `after-search-${Math.floor(Date.now() / 1000)}`,
      { message: `search complete for ${JSON.stringify(input.topic)}` },
    );
    return { sources_count: sources.length, snapshot_id: snap.id };
  });

  runtime.register("research-pipeline", "fetch", async (ctx) => {
    const sourcesIdx = (await ctx.workspace.kv.get("sources/index")) as string[];
    let fetched = 0;
    for (const url of sourcesIdx) {
      const content = mockFetch(url);
      await ctx.workspace.files.write(`sources/${slugify(url)}.txt`, content);
      fetched += 1;
    }
    const snap = await ctx.workspace.snapshot(
      `after-fetch-${Math.floor(Date.now() / 1000)}`,
      { message: `fetched ${fetched} sources` },
    );
    return { fetched_count: fetched, snapshot_id: snap.id };
  });

  // `extract` uses the LLM facade (`ctx.client.llm`) to pull short
  // fact lists from each fetched source. Provider configuration lives
  // in `start-workflow.ts` so the handler stays portable across mock
  // and real providers.
  runtime.register("research-pipeline", "extract", async (ctx) => {
    const topic = (await ctx.workspace.kv.get("topic")) as string;
    const sourcesIdx = (await ctx.workspace.kv.get("sources/index")) as string[];
    const stepInput = (ctx.step.input ?? {}) as { model?: string };
    const model = stepInput.model ?? "claude-sonnet-4-5";

    let extracted = 0;
    let totalInputTokens = 0;
    let totalOutputTokens = 0;
    let totalCostUsd = 0;

    for (const url of sourcesIdx) {
      const path = `sources/${slugify(url)}.txt`;
      let content: string;
      try {
        content = await ctx.workspace.files.readText(path);
      } catch {
        continue;
      }

      const response = await ctx.client.llm.complete({
        model,
        messages: [
          {
            role: "system",
            content:
              "You are a careful research assistant. Extract 3-5 short, " +
              "verifiable facts as a bulleted list. No commentary.",
          },
          {
            role: "user",
            content: `Topic: ${topic}\nSource (${url}):\n${content}`,
          },
        ],
        workspaceId: ctx.workspaceRecord?.id,
      });

      // Persist the raw LLM output so the synth step can read it. The
      // existing synth handler expects `string[]` — we wrap the
      // response in a one-element array to keep the chain working
      // without coupling synth to LLM-shaped output.
      await ctx.workspace.kv.set(`facts/${url}`, [response.content] as unknown as JsonValue);
      totalInputTokens += response.inputTokens;
      totalOutputTokens += response.outputTokens;
      totalCostUsd += response.costUsd;
      extracted += 1;
    }

    const snap = await ctx.workspace.snapshot(
      `after-extract-${Math.floor(Date.now() / 1000)}`,
      { message: `extracted from ${extracted} sources via ${model}` },
    );
    return {
      extracted_count: extracted,
      input_tokens: totalInputTokens,
      output_tokens: totalOutputTokens,
      cost_usd: totalCostUsd,
      model,
      snapshot_id: snap.id,
    };
  });

  runtime.register("research-pipeline", "synth", async (ctx) => {
    const topic = (await ctx.workspace.kv.get("topic")) as string;
    const sourcesIdx = (await ctx.workspace.kv.get("sources/index")) as string[];
    const factsByUrl: Record<string, string[]> = {};
    for (const url of sourcesIdx) {
      try {
        const facts = (await ctx.workspace.kv.get(`facts/${url}`)) as string[];
        if (facts) factsByUrl[url] = facts;
      } catch {
        continue;
      }
    }
    const report = mockSynthesise(topic, factsByUrl);
    await ctx.workspace.files.write("report.md", report);
    const snap = await ctx.workspace.snapshot(
      `after-synth-${Math.floor(Date.now() / 1000)}`,
      { message: "report written" },
    );
    return { report_chars: report.length, snapshot_id: snap.id };
  });
}
